sycl/source/detail/scheduler/commands.cpp

```diff
@@ -2575,8 +2585,17 @@ pi_int32 enqueueImpKernel(
 
   pi_result Error = PI_SUCCESS;
   {
-    assert(KernelMutex);
-    std::lock_guard<std::mutex> Lock(*KernelMutex);
+    struct OptionalLockGuard {
```

Comment:
```
std::unique_lock<std::mutex> Lock(*KernelMutex, std::defer_lock);

if (KernelMutex)
  Lock.lock();

if (Lock.owns_lock()) {
   ...
}
```
?

Reply:
That would dereference `KernelMutex` before checking if it's null, although I believe we can still use `unique_lock`, I think we can do something like this:

```c++
using LockT = std::unique_lock<std::mutex>;
auto lock = m ? LockT(*m) : LockT();
```

And I don't think `.owns_lock()` would be necessary, if the mutex was null, then that just means the caching is disabled  and the kernel can't be shared, we still want to enqueue it.